google/auth/credentials.py

```diff
@@ -191,14 +191,16 @@ def _blocking_refresh(self, request):
             self.refresh(request)
 
     def _non_blocking_refresh(self, request):
-        if (
-            self.token_state == TokenState.STALE
-            and not self._refresh_worker.has_error()
-        ):
-            self._refresh_worker.start_refresh(self, request)
+        use_blocking_refresh_fallback = False
 
-        if self.token_state == TokenState.INVALID:
+        if self.token_state == TokenState.STALE:
+            use_blocking_refresh_fallback = not self._refresh_worker.start_refresh(
+                self, request
+            )
+
+        if self.token_state == TokenState.INVALID or use_blocking_refresh_fallback:
             self.refresh(request)
+            self._refresh_worker.clear_error()
```

Comment:
maybe leave a comment describing why we need to call this right here.

Reply:
Done!